articles/external_qos_file.md

```diff
@@ -0,0 +1,402 @@
+```
+
+### Loader
+
+Function to load the qos profiles file:
+
+```c
+rcl_ret_t
+rcl_qos_load_profiles_from_xml_file(const char * file_path, rcl_qos_loaded_profiles_t * lp);
+```
+
+There could be different functions for different formats.
+
+### Loaded profiles builder
+
+Infrastructure to easily build a `rcl_qos_loaded_profiles_t` object.
+The intent of this API is to decouple and actual parser (e.g. XML), from how a `rcl_qos_loaded_profiles_t` is built.
+
+```c
+/// Default initializer
+rcl_qos_profiles_builder_t
+rcl_get_default_initialized_qos_profiles_builder();
+
+/// Set qos for publisher
+rcl_ret_t
+rcl_qos_profiles_builder_set_publisher_qos(
+    rcl_qos_loaded_profiles_builder_t * builder,
+    const char * node_name,
+    const char * node_ns,
+    const char * topic_name,
+    const rmw_qos_profile_t * qos);
+
+/// Set qos for publisher with profile id
+rcl_ret_t
+rcl_qos_profiles_builder_set_publisher_qos_with_id(
+    rcl_qos_profiles_builder_t * builder,
+    const char * node_name,
+    const char * node_ns,
+    const char * qos_profile_id,
+    const rmw_qos_profile_t * qos);
+
+/// Similar methods for subscriptions/clients/services
+...
+
+/// Build a `rcl_qos_loaded_profiles_t` object
+rcl_ret_t
+rcl_qos_profiles_builder_get_loaded_profiles(rcl_qos_profiles_builder_t * builder, rcl_qos_loaded_profiles_t * lp);
+```
+
+How to parse a `rmw_qos_profile_t` from the qos file is up to the specific loader.
+
+### Usage in init options and node options
+
+A `rcl_qos_loaded_profiles_t` member could be added in both `rcl_init_options_t` and `rcl_node_options_t`.
+The one in init options will be common to all nodes in the same contexts, and can be overridden by the one in node options.
+
+Optionally, we could also add in `rcl_node_options_t` a boolean to indicate to use the `rcl_qos_loaded_profiles_t` in init options if the profile was not found in the node specific loaded profiles object.
+
+### New ros argument
+
+A new argument will be needed to allow the user to specify which file to load.
+
+```bash
+ros2run <package_name> <exec_name> --ros-args --qos-file /path/to/qos/file
+```
+
+This loading mechanism will be equivalent to setting the `rcl_qos_loaded_profiles_t` member in init options.
+
+## Further extensions
+
+### Loading qos profiles in composable nodes
+
+The interface definition of the [LoadNode](https://github.com/ros2/rcl_interfaces/blob/master/composition_interfaces/srv/LoadNode.srv) service could be extended with a field to specify the desired `QoS` of the loaded node.
+
+In that way, the user loading the node is responsible of passing the desired QoS, instead of the node container.
+The message to be sent could be loaded from a qos profile file.
+
+#### Proposed interfaces
+
+```
+# EndpointQoS.msg
+
+string name  # This can be either a topic or service name
+uint8 type  # Endpoint type. An enum that can be a publisher/subscription/topic/service
+QoS qos  # Message that defines the qos
```

Comment:
Do you think a `QoS` message has use without the name and type for the entity?

If not, maybe it's worth moving the `name` and `type` fields to the `QoS` message, leaving us with a single message definition.

Reply:
It might possibly make it easier to reuse instantiated `QoS` messages?

Reply:
I image that a `QoS` message can have value on its own.